src/main/java/com/amazon/opendistroforelasticsearch/sql/optimizer/subquery/InSubqueryRewriter.java

```diff
@@ -0,0 +1,146 @@
+
+/**
+ * IN Subquery Rewriter.
+ * For example,
+ *  SELECT * FROM A WHERE a IN (SELECT b FROM B) and c > 10 should be rewritten to
+ *  SELECT A.* FROM A JOIN B ON A.a = B.b WHERE c > 10 and B.b IS NOT NULL.
+ */
+public class InSubqueryRewriter implements SubqueryRewriter {
+    private final SQLInSubQueryExpr inSubQueryExpr;
+    private final MySqlSelectQueryBlock inSubQuery;
+    private final SQLExpr joinLeftExpr;
+    private final SQLExpr joinRightExpr;
+
+    public InSubqueryRewriter(Subquery subquery) {
+        this.inSubQueryExpr = (SQLInSubQueryExpr) subquery.getSubQueryExpr();
+        this.inSubQuery = (MySqlSelectQueryBlock) inSubQueryExpr.getSubQuery().getQuery();
+        joinLeftExpr = inSubQueryExpr.getExpr();
+        joinRightExpr = rightExpr(inSubQuery);
+    }
+
+    /**
+     * Rewrite the subquery with JOIN clause. This API is NOT idempotent.
+     *
+     * @param query Subquery will be written.
+     */
+    @Override
+    public void rewrite(MySqlSelectQueryBlock query) {
+        query.setFrom(buildFrom(query));
+        query.setWhere(buildWhere(query));
+    }
+
+    /**
+     * Build From clause from input query.
+     *
+     * With the input query.
+     *       Query
+     *     /   |   \
+     * SELECT FROM     WHERE
+     *    |    |     /   |  \
+     *    *    A  c>10 AND INSubquery
+     *                      /    \
+     *                     a    Query
+     *                          /    \
+     *                       SELECT FROM
+     *                         |     |
+     *                         b     B
+     *
+     * The FROM logic should be
+     *     JOIN
+     *     / | \
+     *    A  B A.a == B.b
+     */
+    private SQLJoinTableSource buildFrom(MySqlSelectQueryBlock rootQuery) {
+        SQLJoinTableSource sqlJoin = new SQLJoinTableSource();
+        sqlJoin.setLeft(rootQuery.getFrom());
+        sqlJoin.setRight(inSubQuery.getFrom());
+        sqlJoin.setCondition(new SQLBinaryOpExpr(joinLeftExpr, SQLBinaryOperator.Equality, joinRightExpr));
+        sqlJoin.setJoinType(JoinType.JOIN);
+
+        return sqlJoin;
+    }
+
+    /**
+     * Build Where clause from input query.
+     *
+     * With the input query.
+     *       Query
+     *     /   |   \
+     * SELECT FROM     WHERE
+     *    |    |     /   |  \
+     *    *    A  c>10 AND INSubquery
+     *                      /    \
+     *                     a    Query
+     *                          /    \
+     *                       SELECT FROM
+     *                         |     |
+     *                         b     B
+     *
+     *
+     *     WHERE
+     *     /  |   \
+     *   c>10 AND B.b is NOT NULL
+     */
+    private SQLExpr buildWhere(MySqlSelectQueryBlock rootQuery) {
+        return rewriteSubquery(rootQuery.getWhere());
+    }
+
+    private SQLExpr rewriteSubquery(SQLExpr expr) {
+        if (expr instanceof SQLInSubQueryExpr) {
```

Comment:
np: I still think this is cleaner:

```java
        if (expr instanceof SQLInSubQueryExpr) {
            return createSubqueryReplacementCondition();
        } else {
            if (expr instanceof SQLBinaryOpExpr) {
                SQLBinaryOpExpre binaryExpr = (SQLBinaryOpExpr) expr;
                binaryExpr.setLeft(rewriteSubquery((binaryExpr).getLeft()));
                binaryExpr.setRight(rewriteSubquery((binaryExpr.getRight()));
            }
            return expr;
        }
```

Reply:
Looking into the logic, we can simplify even more

````
if (expr instanceof SQLInSubQueryExpr) {
            return createSubqueryReplacementCondition();
        } 

if (expr instanceof SQLBinaryOpExpr) {
    SQLBinaryOpExpre binaryExpr = (SQLBinaryOpExpr) expr;
    binaryExpr.setLeft(rewriteSubquery((binaryExpr).getLeft()));
    binaryExpr.setRight(rewriteSubquery((binaryExpr.getRight()));
}
return expr;
```